### PR TITLE
Extension of Quail for handling diffusion terms

### PR DESCRIPTION
**Pull request by bornhoft**
This pull request includes the addition of handling second-order terms in DG via the SIP discretization. This includes the following:
- Addition of the SIP diffusion flux in physics/base/functions.py
- Addition of the constant advection diffusion classes (with examples)
- Addition of the Navier-Stokes physics class (with example)
- Addition of interior diffusion fluxes for both classes above
- Addition of tests for NS diffusion flux
- Addition of transport types in navierstokes/tools.py (currently supports constant and Sutherland's law)
- Added the exact flux for linear advection 
- Addition of multiple end-to-end tests
- Removed multiple if statements and cleaned up the main residual functions a bit

I have perhaps missed some of the changes here but I believe we have most things covered. Convergence tests have been completed for 1D/2D constant advection diffusion on both DG/ADER and on quads/tris for DG. Convergence tests have also been completed for 2D manufactured solution for DG (and almost completed for ADERDG).

Please feel free to reach out with more specific questions or comments. I believe that my list includes most of the major changes but this is a reasonably sized pull request so I imagine I have missed a few. Thanks for the extra set of eyes!

### CODE REVIEW

File: test/component/numerics/basis/test_basis.py [hunk 1]
```diff
@@ -33,3 +37,96 @@ def test_lagrange_basis_should_be_nodal(basis, order):
...
+
+@pytest.mark.parametrize('order', [
+	# Order of Lagrange basis
+	0, 1, 2, 3, 4, 5,
+])
+@pytest.mark.parametrize('Basis', [
+	# Basis class representing the element geometry
+	basis_defs.LegendreSeg, basis_defs.LegendreQuad
+])
+def test_legengre_massmatrix_should_be_diagonal(basis, order):
```
Comment (alilasemi) [inline test/component/numerics/basis/test_basis.py:50]: Legendre spelling

File: test/component/physics/navierstokes/test_navierstokes.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		+ rho * v * dvdy + rho * R / (gamma - 1.) * dTdy
+
+	Fref = np.zeros([1, 1, ns, 2])
+	Fref[:, :, irhou, 0] = tauxx
+	Fref[:, :, irhov, 0] = tauxy
+	Fref[:, :, irhoE, 0] = tauxx * u + tauxy * v + kappa * dTdx
+
+	Fref[:, :, irhou, 1] = tauxy
+	Fref[:, :, irhov, 1] = tauyy
+	Fref[:, :, irhoE, 1] = tauxy * u + tauyy * v + kappa * dTdy
```
Comment (alilasemi) [inline test/component/physics/navierstokes/test_navierstokes.py:88]: Try to think of more edge-case tests, as opposed to this, since this basically looks like a re-implementation of the actual flux code. Think of cases like, what if there are only gradients in one direction? Then set the gradient to an actual number (using random numbers is tricky because it's harder to hand calculate). Think of what this flux would be for small hand-calculatable problems like couette flow or something (incompressible so drdx and drdy are both zero, only dudy is nonzero, therefore tauxx must be zero, etc etc).

File: test/end_to_end/cases/scalar/2D/constant_advection_diffusion/ader/input_file.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+TimeStepping = {
+	"InitialTime" : 0.,
+	"FinalTime" : 0.5,
+	# "CFL" : 0.0025,
```
Comment (alilasemi) [inline test/end_to_end/cases/scalar/2D/constant_advection_diffusion/ader/input_file.py:4]: Commented code

File: test/end_to_end/conftest.py [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
 import numpy as np
 import pytest
+import sys
 
 import list_of_cases
 
 
 @pytest.fixture
-def test_data():
+def test_data(capsys):
```
Comment (alilasemi) [inline test/end_to_end/conftest.py:10]: You can remove the capsys. Sorry this is actually something I added lol. Not sure why it shows up here. Anyways it's not needed anymore

**Review by alilasemi — CHANGES_REQUESTED** [review]
Wow this was a lot of stuff, good job! Looks pretty clean and good overall. Just added some comments here and there to address. Main overall comment is to try to split unrelated things into multiple PR's in the future (for example, circleCI/codecov stuff probably couldve been a separate PR, and maybe separate PR's for navier stokes physics vs. navier stokes ADER stuff). Otherwise this looks good!

Some other comments, since apparently Github doesn't let you comment on binary files:
you've include  the following binary data files:
examples/scalar/2D/constant_advection/convergence_testing/RK4/P1.pkl
examples/scalar/2D/constant_advection/convergence_testing/RK4/P2.pkl
examples/scalar/2D/constant_advection/convergence_testing/RK4/P3.pkl
examples/scalar/2D/constant_advection/convergence_testing/RK4/P4.pkl
Why is this? In general we don't want to be storing data in version control.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            pytest --cov=../src --cov-report=xml
+            cd end_to_end
+            python generate_tests.py
+            cd ..
+            pytest -v
+
+#       - run:
+#           name: Send to codecov
+#           command: |
+#             bash <(curl -s https://codecov.io/bash) -Z
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply .circleci/config.yml:57]: Added additional comments

File: .circleci/environment.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  - sqlite=3.36.0=hce871da_0
+  - tk=8.6.11=h7bc2e8c_0
+  - toml=0.10.2=pyhd3eb1b0_0
+  - tornado=6.1=py39h9ed2024_0
+  - tzdata=2021e=hda174b7_0
+  - wheel=0.37.0=pyhd3eb1b0_1
+  - xz=5.2.5=h1de35cc_0
+  - zlib=1.2.11=h1de35cc_3
+  - zstd=1.4.9=h322a384_0
+prefix: /Users/brettbornhoft/opt/anaconda2/envs/quail
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply .circleci/environment.yml:62]: This file has been removed

File: .circleci/config.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (bornhoft) [reply .circleci/config.yml:57]: Removed commented code

File: .circleci/environment.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (bornhoft) [reply .circleci/environment.yml:62]: Removed this file

File: examples/euler/1D/moving_shock/moving_shock.py [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 TimeStepping = {
 	"InitialTime" : 0.,
-	"FinalTime" : 2.e-4,
+	"FinalTime" : 2.e-5,
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/euler/1D/moving_shock/moving_shock.py:5]: Reverted back

File: examples/euler/1D/moving_shock/moving_shock.py [hunk 8]
```diff
@@ -12,7 +12,9 @@
 	"SolutionBasis" : "LagrangeSeg",
 	"Solver" : "DG",
 	"L2InitialCondition" : True,
-	"ApplyLimiters" : "PositivityPreserving",
+	"ApplyLimiters" : "WENO",#"PositivityPreserving",
+	"ShockIndicator" : "MinMod",
+
 }
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/euler/1D/moving_shock/moving_shock.py:18]: Reverted back

File: examples/navierstokes/2D/manufactured_solution/convergence_testing/process.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
+'''
+Script that takes in solution files for various time integration schemes and
+plots their order of convergence (error vs time)
+
+Author: Brett Bornhoft
+Date: 09/27/2021
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/manufactured_solution/convergence_testing/process.py:6]: But I really need the recognition

File: test/component/numerics/basis/test_ader_tools.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+	elif order == 2:
+		expected = 4.0 * np.array([[1., 0., 0., 0., 0., 0., 0., 0., 0.], 
+				 	   	  [0., 1./3., 0., 0., 0., 0., 0., 0., 0.],
+					      [0., 0., 1./5., 0., 0., 0., 0., 0., 0.],
+					      [0., 0., 0., 1./3., 0., 0., 0., 0., 0.],
+					      [0., 0., 0., 0., 1./9., 0., 0., 0., 0.],
+					      [0., 0., 0., 0., 0., 1./15., 0., 0., 0.],
+					      [0., 0., 0., 0., 0., 0., 1./5, 0., 0.],
+					      [0., 0., 0., 0., 0., 0., 0., 1./15., 0.],
+					      [0., 0., 0., 0., 0., 0., 0., 0., 1./25.]])
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply test/component/numerics/basis/test_ader_tools.py:51]: fixed

File: test/component/numerics/basis/test_basis.py (hunk 1 above)
Comment (bornhoft) [reply test/component/numerics/basis/test_basis.py:50]: fixed

File: test/component/physics/navierstokes/test_navierstokes.py (hunk 2 above)
Comment (bornhoft) [reply test/component/physics/navierstokes/test_navierstokes.py:88]: I think adding a few of these tests in the future is a good idea, but I still think this one is good as it doesn't do the product rules for tauxx, tauxy, etc... It just calculates the equations as they are written and verifies that it matches our implementation. We also have the zero velocity cases as well to assist with this function.

File: examples/navierstokes/2D/manufactured_solution/grad.cpp [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+gUy[3] = (4.0*(-0.29999999999999999*M_PI*sin(M_PI*x2)*cos(M_PI*x1) - 0.29999999999999999*M_PI*sin(M_PI*x2))*(0.14999999999999999*sin(3*M_PI*x1) + 0.14999999999999999*cos(M_PI*x1)*cos(M_PI*x2) + 0.14999999999999999*cos(M_PI*x2) + 1) + 4.0*(-0.29999999999999999*M_PI*sin(M_PI*x2)*cos(M_PI*x1) + 0.29999999999999999*M_PI*cos(M_PI*x2))*(0.14999999999999999*sin(M_PI*x2) + 0.14999999999999999*cos(M_PI*x1)*cos(M_PI*x2) + 0.14999999999999999*cos(M_PI*x1) + 1))*(0.050000000000000003*sin(M_PI*x1) + 0.050000000000000003*cos(M_PI*x1)*cos(M_PI*x2) - 0.10000000000000001*cos(M_PI*x2) + 0.5) + (-0.050000000000000003*M_PI*sin(M_PI*x2)*cos(M_PI*x1) + 0.10000000000000001*M_PI*sin(M_PI*x2))*(4.0*pow(0.14999999999999999*sin(3*M_PI*x1) + 0.14999999999999999*cos(M_PI*x1)*cos(M_PI*x2) + 0.14999999999999999*cos(M_PI*x2) + 1, 2) + 4.0*pow(0.14999999999999999*sin(M_PI*x2) + 0.14999999999999999*cos(M_PI*x1)*cos(M_PI*x2) + 0.14999999999999999*cos(M_PI*x1) + 1, 2)) + (-0.5*M_PI*sin(M_PI*x2)*cos(M_PI*x1) + 1.0*M_PI*cos(M_PI*x2))/(gamma - 1);
+
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/manufactured_solution/grad.cpp:16]: removed

File: examples/navierstokes/2D/manufactured_solution/init.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+Uq[irho] = 0.1*np.sin(np.pi*x1) + 0.1*np.cos(np.pi*x1)*np.cos(np.pi*x2) - 0.2*np.cos(np.pi*x2) + 1.0
+
+Uq[irhou] = (0.1*np.sin(np.pi*x1) + 0.1*np.cos(np.pi*x1)*np.cos(np.pi*x2) - 0.2*np.cos(np.pi*x2) + 1.0)*(0.3*np.sin(3*np.pi*x1) + 0.3*np.cos(np.pi*x1)*np.cos(np.pi*x2) + 0.3*np.cos(np.pi*x2) + 2.0)
+
+Uq[irhov] = (0.1*np.sin(np.pi*x1) + 0.1*np.cos(np.pi*x1)*np.cos(np.pi*x2) - 0.2*np.cos(np.pi*x2) + 1.0)*(0.3*np.sin(np.pi*x2) + 0.3*np.cos(np.pi*x1)*np.cos(np.pi*x2) + 0.3*np.cos(np.pi*x1) + 2.0)
+
+Uq[irhoE] = (4.0*(0.15*np.sin(3*np.pi*x1) + 0.15*np.cos(np.pi*x1)*np.cos(np.pi*x2) + 0.15*np.cos(np.pi*x2) + 1)**2 + 4.0*(0.15*np.sin(np.pi*x2) + 0.15*np.cos(np.pi*x1)*np.cos(np.pi*x2) + 0.15*np.cos(np.pi*x1) + 1)**2)*(0.05*np.sin(np.pi*x1) + 0.05*np.cos(np.pi*x1)*np.cos(np.pi*x2) - 0.1*np.cos(np.pi*x2) + 0.5) + (1.0*np.sin(np.pi*x2) + 0.5*np.cos(np.pi*x1)*np.cos(np.pi*x2) + 1.0*np.cos(2*np.pi*x1) + 10.0)/(gamma - 1)
+
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/manufactured_solution/init.py:8]: removed

File: examples/navierstokes/2D/manufactured_solution/manufactured_NS_source.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+#code = ccode(Assignment(gUx_rho, rho_x)) + '\n\n'
+#code += ccode(Assignment(gUx_rhou, rhou_x)) + '\n\n'
+#code += ccode(Assignment(gUx_rhov, rhov_x)) + '\n\n'
+#code += ccode(Assignment(gUx_rhoE, rhoE_x)) + '\n\n'
+#code += ccode(Assignment(gUy_rho, rho_y)) + '\n\n'
+#code += ccode(Assignment(gUy_rhou, rhou_y)) + '\n\n'
+#code += ccode(Assignment(gUy_rhov, rhov_y)) + '\n\n'
+#code += ccode(Assignment(gUy_rhoE, rhoE_y)) + '\n\n'
+#with open('grad.cpp', 'w') as f:
+#    f.write(code)
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/manufactured_solution/manufactured_NS_source.py:112]: removed

File: examples/navierstokes/2D/manufactured_solution/script_convergence.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+def print_errors(N, errors):
+	for i in range(errors.shape[0]-1):
+		err = np.log(errors[i+1]/errors[i]) / np.log(N[i+1]/N[i])
+		print(err)
+
+def array_errors(N, errors):
+	err=np.zeros([N.shape[0]-1])
+	for i in range(errors.shape[0]-1):
+		err[i] = np.log(errors[i+1]/errors[i]) / np.log(N[i+1]/N[i])
+	return err
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/manufactured_solution/script_convergence.py:23]: cleaned

File: examples/navierstokes/2D/manufactured_solution/source.py [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/manufactured_solution/source.py:8]: removed

File: examples/navierstokes/2D/taylorgreenvortex/create_anim.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+	# Unpack
+	mesh = solver.mesh
+	physics = solver.physics
+
+	# Plot solution
+	if j == 0:
+		ignore_colorbar = False
+	else:
+		ignore_colorbar = True
+	#levels = np.array([0.45, 0.6, 0.75, 0.9, 1.05, 1.2, 1.35, 1.5])
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/taylorgreenvortex/create_anim.py:35]: removed file as its not needed

File: examples/navierstokes/2D/taylorgreenvortex/post_process.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+fname = "Data_final.pkl"
+solver = readwritedatafiles.read_data_file(fname)
+
+# Unpack
+mesh = solver.mesh
+physics = solver.physics
+
+# Compute L2 error
+post.get_error(mesh, physics, solver, "Pressure", normalize_by_volume=False)
+#levels = np.array([0.45, 0.6, 0.75, 0.9, 1.05, 1.2, 1.35, 1.5])
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/navierstokes/2D/taylorgreenvortex/post_process.py:16]: removed

File: examples/scalar/1D/constant_advection_diffusion/convergence.py [hunk 18]
```diff
@@ -0,0 +1,181 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply examples/scalar/1D/constant_advection_diffusion/convergence.py:1]: removed

File: src/numerics/timestepping/stepper.py [hunk 19]
```diff
@@ -277,15 +277,18 @@ def take_time_step(self, solver):
 		Time = solver.time
 		for istage in range(self.nstages):
 			dt = self.dt
-			solver.time = Time + dt
+			#HACK -> Need to confirm that this is the right thing to do!
+
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/numerics/timestepping/stepper.py:281]: In main we have the solver.time attribute updated prior to the get_residual and mult_inv_mass_matrix function calls. This becomes an issue for time dependent source terms. We were previously evaluating any source term at the next time step rather than the current one. This never showed up before because we rarely have time dependent source terms.

File: src/physics/base/functions.py [hunk 20]
```diff
@@ -188,3 +196,206 @@ def compute_flux(self, physics, UqL, UqR, normals):
...
+		# Calculate ratio of volume/area for each L/R face
+		self.hL = vol_elems[int_face_helpers.elemL_IDs] / \
+				face_lengths[int_face_helpers.elemL_IDs, -1]
+		
+		'''
+		HACK / Note: This needs investigation... needed to subtract one for 
+		1D cases with BCs to get the ID mapping correct
+		Not sure if this is general for 2D... (cases all converge but 
+		perhaps some edge cases would cause issues.)
+		'''
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/physics/base/functions.py:241]: This has been verified.

File: src/physics/euler/functions.py [hunk 21]
```diff
@@ -1263,7 +1263,15 @@ def compute_flux(self, physics, UqL_std, UqR_std, normals):
...
-
+		
+		# Entropy fix
+		# eps = np.zeros_like(evals)
+		# eps[:, :, :] = (1e-2 * c)
+		# fix = np.argwhere(np.logical_and(evals < eps, evals > -eps))
+		# fix_shape = fix[:, 0], fix[:, 1], fix[:, 2]
+		# evals[fix_shape] = 0.5 * (eps[fix_shape] + evals[fix_shape]* \
+		# 	evals[fix_shape] / eps[fix_shape])
+
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/physics/euler/functions.py:1274]: Keeping this for now as it is documented.

File: src/physics/navierstokes/functions.py [hunk 22]
```diff
@@ -0,0 +1,465 @@
...
+	TaylorGreenVortexNS = auto()
+	ManufacturedSolution = auto()
+
+# class BCType(Enum):
+	'''
+	Enum class that stores the types of boundary conditions. These
+	boundary conditions are specific to the available Euler equation sets.
+	'''
+	# SlipWall = auto()
+	# PressureOutlet = auto()
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/physics/navierstokes/functions.py:34]: removed and added a pass for now to indicate that future bcs can be added.

File: src/physics/navierstokes/functions.py [hunk 23]
```diff
@@ -0,0 +1,465 @@
...
+			t, gamma, kappa, 
+			mu, R)
+
+		
+		return Sq # [ne, nq, ns]
+
+	def manufactured_source(self, x1, x2, t, gamma, kappa, mu, R):
+	
+		
+		# The following lines of code are generated using sympy
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/physics/navierstokes/functions.py:202]: Yes, but I think the infrastructure for doing that isn't quite in place. We can add this as a future issue if you are fine with that.

File: src/physics/navierstokes/navierstokes.py [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+			varq = np.linalg.norm(mom, axis=2, keepdims=True)/rho + np.sqrt(
+					gamma*get_pressure()/rho)
+		elif vname is self.AdditionalVariables["Velocity"].name:
+			varq = np.linalg.norm(mom, axis=2, keepdims=True)/rho 
+		elif vname is self.AdditionalVariables["XVelocity"].name:
+			varq = mom[:, :, [0]]/rho
+		elif vname is self.AdditionalVariables["YVelocity"].name:
+			varq = mom[:, :, [1]]/rho
+		else:
+			raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/physics/navierstokes/navierstokes.py:165]: Yes, this has been condensed.

File: src/solver/DG.py [hunk 25]
```diff
@@ -267,6 +273,30 @@ class InteriorFaceHelpers(ElemHelpers):
...
+		stores the evaluated determinant of the geometric Jacobian for
+		each left element
+	jacR_elems: numpy array
+		stores the evaluated geometric Jacobian for each right element
+	ijacR_elems: numpy array
+		stores the evaluated inverse of the geometric Jacobian for each
+		right element
+	djacR_elems: numpy array
+		stores the evaluated determinant of the geometric Jacobian for
+		each right element
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/solver/DG.py:299]: We only need ijacR/L_elems... others have been removed.
ijacR/L is used to convert the reference gradient of the state to the physical gradient of the state at the face quadrature points.

[the remark this answers is not in the document]
Comment (bornhoft) [reply src/solver/DG.py:299]: We only need ijacR/L_elems... others have been removed.
ijacR/L is used to convert the reference gradient of the state to the physical gradient of the state at the face quadrature points.

File: src/solver/ader_tools.py [hunk 26]
```diff
@@ -446,6 +493,62 @@ def calculate_source_term_integral(elem_helpers, elem_helpers_st, Sq):
...
+	Outpus:
+	-------
+		gUc: gradient of the space-time state [ne, nb_st, ns, ndims]
+	'''
+	ader_helpers = solver.ader_helpers
+
+	iMM = ader_helpers.iMM
+	SMS = ader_helpers.SMS_ref
+	
+	# SMSxUc = np.einsum('mnl, imk -> inkl', SMS, Uc)
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/solver/ader_tools.py:514]: removed

File: src/solver/ader_tools.py [hunk 27]
```diff
@@ -729,6 +836,9 @@ def predictor_elem_stiffimplicit(solver, dt, W, U_pred):
 	# Initialize space-time coefficients
 	U_pred, U_bar = solver.get_spacetime_guess(solver, W, U_pred, dt=dt)
 
+	# Calculate the gradient of the state
+	# gU_pred = get_spacetime_gradient(solver, U_pred)
+
```
[the remark this answers is not in the document]
Comment (bornhoft) [reply src/solver/ader_tools.py:841]: moved this into flux_coefficients and removed from this file

**Review by alilasemi — APPROVED** [review]
